kyo-kernel/shared/src/test/scala/kyo/kernel/ArrowEffectTest.scala

```diff
@@ -337,4 +337,210 @@ class ArrowEffectTest extends Test:
+                        assert(v == (42 * 10) + ("a".size * 10))
+                    }
+            }
+
+            "multiple shift with different effect sets" in run {
+                Delim.shift[Int, Int, TestEffect2] { k =>
+                    k(42).map { r =>
+                        testEffect2("a").map(v => r + v)
+                    }
+                }.map(_ * 10).map { v1 =>
+                    Delim.shift[Int, Int, TestEffect1] { k =>
+                        k(42).map { r =>
+                            testEffect1(v1).map(s => r + s.length)
+                        }
+                    }
+                }
+                    .handle(
+                        Delim.run,
+                        ArrowEffect.handle(Tag[TestEffect1], _) {
+                            [C] => (input, cont) => cont(input.toString)
+                        },
+                        ArrowEffect.handle(Tag[TestEffect2], _) {
+                            [C] => (input, cont) => cont(input.size)
+                        }
+                    ).map { v =>
+                        assert(v == 46)
+                    }
+            }
+
+            "short circuiting" in run {
+                def test(numbers: List[Int], expected: Int) =
+                    Kyo.foldLeft(numbers)(0) { (acc, n) =>
+                        if n < 0 || n == 42 then
+                            Delim.shift[Int, Int, Any] { _ => -1 }
+                        else
+                            acc + n
+                    }
+                        .handle(Delim.run)
+                        .map { r =>
+                            assert(r == expected)
+                        }
+
+                for
+                    _ <- test(List(1, 2), 3)
+                    _ <- test(List(1, 2, -1), -1)
+                    _ <- test(List(1, 2, 42, 3), -1)
+                yield succeed
+                end for
+            }
+        }
+
+        "flow effect with dynamic tags" - {
+            sealed trait Flow[+In, -Out] extends ArrowEffect[Flow.Op[In, Out, *], Id]
+
+            object Flow:
+                enum Op[-In, +Out, R]:
+                    case Poll[V]()     extends Op[V, Nothing, Maybe[V]]
+                    case Emit[V](v: V) extends Op[Any, V, Unit]
+
+                def emit[V: Tag](value: V): Unit < Flow[Any, V] =
+                    ArrowEffect.suspend(Tag[Flow[Any, V]], Op.Emit(value))
+
+                def poll[V: Tag]: Maybe[V] < Flow[V, Nothing] =
+                    ArrowEffect.suspend(Tag[Flow[V, Nothing]], Op.Poll())
+
+                def run[A, S, In: Tag, Out: Tag](in: Chunk[In])(v: A < (Flow[In, Out] & S)): (Chunk[In], Chunk[Out], A) < S =
+                    ArrowEffect.handleLoop(Tag[Flow[In, Out]], (in, Chunk.empty[Out]), v)(
+                        [C] =>
+                            (input, state, cont) =>
+                                val (in, out) = state
+                                (input: @unchecked) match
```

Comment:
The need for `unchecked` here seems a compiler bug. It warns that `case Op.Emit(v)` is unreachable but that's not the case.